pkg/kubectl/history.go

```diff
@@ -47,38 +47,57 @@ const (
 	ChangeCauseAnnotation = "kubernetes.io/change-cause"
 )
 
+//this will store intermediate state to be later used to visit given element and fetch it's history
+type visitorHistoryViewer struct {
+	element schema.GroupKind
+	c       clientset.Interface
+}
+
 // HistoryViewer provides an interface for resources have historical information.
 type HistoryViewer interface {
 	ViewHistory(namespace, name string, revision int64) (string, error)
 }
 
 func HistoryViewerFor(kind schema.GroupKind, c clientset.Interface) (HistoryViewer, error) {
-	switch kind {
```

Comment:
I would minimize the code changes, and keep the viewers.  Just use the Visitor to pick which one to return.

Reply:
So if I understand correctly, I should keep `DeploymentHistoryViewer, ...` and use Visitor only inside this function to decide which one of them to return. But the visitor pattern in `kind_visitor.go` has only an `Accept` method which does not return anything.

Reply:
Yes, start with changing the fewest lines of code to eliminate the switch statement and referenced api packages.